File: terra/src/main/scala/com/dafttech/terra/game/world/entities/EntityItem.java
package com.dafttech.terra.game.world.entities;

import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.dafttech.terra.engine.Vector2;
import com.dafttech.terra.game.world.World;
import com.dafttech.terra.game.world.items.Item;
import com.dafttech.terra.game.world.items.inventories.Stack;
import com.dafttech.terra.game.world.items.persistence.Prototype;
import com.dafttech.terra.resources.Options;

public class EntityItem extends Entity {
    Item wrapped;

    public EntityItem(Vector2 pos, World world, Vector2 size, Item w) {
        super(pos, world, size);
        wrapped = w;
    }

    @Override
    public TextureRegion getImage() {
        return wrapped.getImage();
    }

    @Override
    public Prototype toPrototype() {
        return wrapped.toPrototype();
    }

    @Override
    public boolean collidesWith(Entity e) {
        return false;
    }

    @Override
    public void update(World world, float delta) {
        Vector2 p = getWorld().localPlayer().getPosition();
        Vector2 s = getWorld().localPlayer().getSize();

        Vector2 vp = p.$plus(s.x() * Options.BLOCK_SIZE() / 2, s.y() * Options.BLOCK_SIZE() / 2);
        vp.$minus(getPosition());

        if (vp.length$u00B2() < 400) {
            getWorld().localPlayer().inventory().add(Stack.apply(wrapped, 1));

            getWorld().removeEntity(this);
        }

        if (vp.length$u00B2() < 10000) {
            addForce(vp.normalized().$times(14f / vp.length$u00B2()));
        }

        super.update(world, delta);
    }
}

File: terra/src/main/scala/com/dafttech/terra/game/world/entities/EntityItem.scala
package com.dafttech.terra.game.world.entities

import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.dafttech.terra.engine.TilePosition
import com.dafttech.terra.engine.vector.Vector2d
import com.dafttech.terra.game.world.GameWorld
import com.dafttech.terra.game.world.items.Item
import com.dafttech.terra.game.world.items.inventories.Stack
import com.dafttech.terra.game.world.items.persistence.Prototype
import com.dafttech.terra.resources.Options
import monix.eval.Task

class EntityItem(pos: Vector2d, val wrapped: Item)(implicit world: GameWorld) extends Entity(pos) {
  override def getImage: Task[TextureRegion] = wrapped.getImage

  override def toPrototype: Prototype = wrapped.toPrototype

  override def collidesWith(e: Entity): Boolean = false

  override def update(delta: Float)(implicit tilePosition: TilePosition): Unit = {
    val vp = getWorld.localPlayer.getPosition - getPosition

    if (vp.`length²` < 400) {
      getWorld.localPlayer.inventory.add(Stack.apply(wrapped, 1))
      getWorld.removeEntity(this)
    }

    val force = vp.normalized * (14f / vp.`length²`)

    if (vp.`length²` < 10000) body.applyForceToCenter(force.x.toFloat, force.y.toFloat, true)
    super.update(delta)
  }
}